src/helpers/transport/mod.rs

```diff
@@ -1,114 +1,196 @@
+        NoResourceIdentifier
     }
-}
 
-/// The source of the command, i.e. where it came from. Some may arrive from helper peers, others
-/// may come directly from the clients
-#[derive(Debug, Eq, PartialEq)]
-pub enum CommandOrigin {
-    Helper(HelperIdentity),
-    Other,
+    fn query_id(&self) -> QueryId {
+        self.0
+    }
+
+    fn step(&self) -> Step {
+        self.1.clone()
+    }
+
+    fn extra(&self) -> &str {
+        ""
+    }
 }
 
-/// Wrapper around `TransportCommand` that indicates the origin of it.
-#[derive(Debug)]
-pub struct CommandEnvelope {
-    pub origin: CommandOrigin,
-    pub payload: TransportCommand,
+impl RouteParams<RouteId, QueryId, Step> for (RouteId, QueryId, Step) {
+    fn resource_identifier(&self) -> RouteId {
+        self.0
+    }
+
+    fn query_id(&self) -> QueryId {
+        self.1
+    }
+
+    fn step(&self) -> Step {
+        self.2.clone()
+    }
+
+    fn extra(&self) -> &str {
+        ""
+    }
 }
 
-/// Represents the transport layer of the IPA network. Allows layers above to subscribe for events
-/// arriving from helper peers or other parties (clients) and also reliably deliver messages using
-/// `send` method.
+/// Transport that supports per-query,per-step channels
 #[async_trait]
-pub trait Transport: Send + Sync + 'static {
-    type CommandStream: Stream<Item = CommandEnvelope> + Send + Sync + Unpin;
+pub trait Transport: Clone + Send + Sync + 'static {
+    type RecordsStream: Stream<Item = Vec<u8>> + Send + Unpin;
 
-    /// Returns the identity of the helper that runs this transport
     fn identity(&self) -> HelperIdentity;
 
-    /// To be called by an entity which will handle the events as indicated by the
-    /// [`SubscriptionType`]. There should be only 1 subscriber per type.
-    /// # Panics
-    /// May panic if attempt to subscribe to the same [`SubscriptionType`] twice
-    async fn subscribe(&self, subscription: SubscriptionType) -> Self::CommandStream;
+    async fn send<D, Q, S, R>(
+        &self,
+        dest: HelperIdentity,
+        route: R,
+        data: D,
+    ) -> Result<(), io::Error>
+    where
+        Option<QueryId>: From<Q>,
+        Option<Step>: From<S>,
+        Q: QueryIdBinding,
+        S: StepBinding,
+        R: RouteParams<RouteId, Q, S>,
+        D: Stream<Item = Vec<u8>> + Send + 'static;
 
-    /// To be called when an entity wants to send commands to the `Transport`.
-    async fn send<C: Send + Into<TransportCommand>>(
+    /// Return the stream of records to be received from another helper for the specific query
+    /// and step
+    fn receive<R: RouteParams<NoResourceIdentifier, QueryId, Step>>(
         &self,
-        destination: HelperIdentity,
-        command: C,
-    ) -> Result<(), TransportError>;
+        from: HelperIdentity,
+        route: R,
+    ) -> Self::RecordsStream;
+}
+
+/// Enum to dispatch calls to various [`Transport`] implementations without the need
```

Comment:
A couple possibilities:
* If the choice is always known at compile time, use type aliases
* Make transport object-safe. There is a `dyn-clone` crate to support cloning (if we only ever have boxed instances and not references). The routing type parameters could probably be eliminated, either by taking an `Addr` instead, or serializing the route information, or some combination.

Reply:
I think transport is actually object-safe, associated type is the problem here. I can make `Box<dyn T>>` but I have to constrain associated type as well, otherwise it won't compile. 

Looking at `TransportImpl` again, it basically implements type alias approach with quirks